validate -u fails

### Checked for duplicates

No - I haven't checked

### 🐛 Describe the bug

validate -u ERRORs out instead of updating the resource file

### 🕵️ Expected behavior

I expected [...]

### 📜 To Reproduce

From my mac:
% validate -u
[snip]
  FAIL: file:/Users/rchen/PDS4tools/validate/resources/registered_context_products.json
      ERROR  [error.connection.registry]   Error connecting to Registry to update registered context products config file. Verify internet connection and try again.
        1 product validation(s) completed
java.io.IOException: Server returned HTTP response code: 400 for URL: https://pds.mcp.nasa.gov/api/search/1/products?limit=1000&sort=ops:Harvest_Info.ops:harvest_date_time&q=%28product_class+eq+%22Product_Context%22+and+pds%3AResource.pds%3Atype+ne+%22Information.Science_Portal%22+and+pds%3AResource.pds%3Atype+ne+%22Information.Investigation%22%29
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:2014)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1615)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:223)
	at java.base/java.net.URL.openStream(URL.java:1325)
	at gov.nasa.pds.validate.ValidateLauncher.getLatestJsonContext(ValidateLauncher.java:552)
	at gov.nasa.pds.validate.ValidateLauncher.processMain(ValidateLauncher.java:1809)
	at gov.nasa.pds.validate.ValidateLauncher.main(ValidateLauncher.java:1888)
java.lang.NullPointerException: Cannot invoke "java.util.Scanner.close()" because "reader" is null


### 🖥 Environment Info

%validate --version
gov.nasa.pds:validate
Version 4.2.0
Release Date: 2026-06-15 22:25:06

I'm pretty confident the ingest to OpenSearch worked. I added 2 new LIDs, and here's one:
(base) [pds4@cds-prod-gamma ~]$ LIDVID=urn:nasa:pds:context:target:asteroid.162421_2000_et70::1.0
(base) [pds4@cds-prod-gamma ~]$ curl --get "https://pds.mcp.nasa.gov/api/search/1/products/$LIDVID"
{"id":"urn:nasa:pds:context:target:asteroid.162421_2000_et70::1.0","type":"Product_Context","title":"(162421) 2000 ET70","investigations":[],"observing_system_components":[],"targets":[], [snip]

although the ingest log had errors I did not understand:
(base) [pds4@cds-prod-gamma ~]$ grep ERROR log/harvest.context_pds4-20260819110135.log 
2026-08-19 11:01:55 [ERROR] (LddEsJsonWriter.java:writeRecord:141) The field 'directory_path_name' will not be searchable because LDD does not contain a type for it (see harvest#204).
2026-08-19 11:01:58 [ERROR] (SchemaUpdater.java:updateSchema:131) Could not find the data type for the field pds:Product_Context/pds:Reference_List
2026-08-19 11:01:58 [ERROR] (FilesProcessor.java:onFile:172) Could not find datatype(s). See 'https://nasa-pds.github.io/registry/user/load1.html' for more information.
2026-08-19 11:02:02 [ERROR] (SchemaUpdater.java:updateSchema:131) Could not find the data type for the field pds:Product_Context/pds:Discipline_Area
2026-08-19 11:02:02 [ERROR] (FilesProcessor.java:onFile:172) Could not find datatype(s). See 'https://nasa-pds.github.io/registry/user/load1.html' for more information.
[snip][


### 📚 Version of Software Used

_No response_

### 🩺 Test Data / Additional context

_No response_

### 🦄 Related requirements

🦄 #xyz

### ⚙️ Engineering Details

_No response_

### 🎉 Integration & Test

_No response_

--- a/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
+++ b/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
@@ -532,7 +532,7 @@ public class ValidateLauncher {
   private void getLatestJsonContext() {
     final String searchAfterParam = "search-after";
     final int pageSize = 1000;
-    final String searchAfterKey = "ops:Harvest_Info.ops:harvest_date_time";
+    final String searchAfterKey = "ops:Label_File_Info.ops:creation_date_time";
     List<ValidationProblem> pList = new ArrayList<>();
     ObjectMapper mapper = new ObjectMapper();
     String base = ToolInfo.getSearchURL();

--- a/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
+++ b/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
@@ -532,7 +532,7 @@ public class ValidateLauncher {
   void getLatestJsonContext() {
     final String searchAfterParam = "search-after";
     final int pageSize = 1000;
-    final String searchAfterKey = "ops:Label_File_Info.ops:creation_date_time";
+    final String sortKey = "ops:Label_File_Info.ops:creation_date_time,lidvid";
     List<ValidationProblem> pList = new ArrayList<>();
     ObjectMapper mapper = new ObjectMapper();
     String base = ToolInfo.getSearchURL();
@@ -546,7 +546,7 @@ public class ValidateLauncher {
       List<Map<String, Object>> contexts = new ArrayList<Map<String, Object>>();
       do {
         url = new URL(base + "/" + endpoint + "?limit=" + Integer.toString(pageSize) + "&q="
-            + URLEncoder.encode(query, StandardCharsets.UTF_8) + "&sort=" + searchAfterKey + "&"
+            + URLEncoder.encode(query, StandardCharsets.UTF_8) + "&sort=" + sortKey + "&"
             + searchAfter);
         LOG.debug("Query URL: " + url.toString());
         reader = new Scanner(url.openStream()).useDelimiter("\\Z");
@@ -559,8 +559,8 @@ public class ValidateLauncher {
         List<Map<String, Object>> dataDocuments = (List<Map<String, Object>>) response.get("data");
 
         contexts.addAll(dataDocuments);
-        String searchAfterValue =
-            getSearchAfterFromDocument(dataDocuments.get(dataDocuments.size() - 1), searchAfterKey);
+        Map<String, Object> lastDoc = dataDocuments.get(dataDocuments.size() - 1);
+        String searchAfterValue = getSearchAfterFromDocument(lastDoc);
 
         searchAfter =
             searchAfterParam + "=" + URLEncoder.encode(searchAfterValue, StandardCharsets.UTF_8);
@@ -606,10 +606,13 @@ public class ValidateLauncher {
     }
   }
 
-  private String getSearchAfterFromDocument(Map<String, Object> document, String searchAfterKey) {
-    @SuppressWarnings("unchecked")
+  @SuppressWarnings("unchecked")
+  private String getSearchAfterFromDocument(Map<String, Object> document) {
     Map<String, Object> properties = (Map<String, Object>) document.get("properties");
-    return ((List<String>) properties.get(searchAfterKey)).get(0);
+    String creationDateTime =
+        ((List<String>) properties.get("ops:Label_File_Info.ops:creation_date_time")).get(0);
+    String lidvid = ((List<String>) properties.get("lidvid")).get(0);
+    return creationDateTime + "," + lidvid;
   }
 
   @SuppressWarnings("unchecked")

--- a/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
+++ b/src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
@@ -529,7 +529,7 @@ public class ValidateLauncher {
     }
   }
 
-  private void getLatestJsonContext() {
+  void getLatestJsonContext() {
     final String searchAfterParam = "search-after";
     final int pageSize = 1000;
     final String searchAfterKey = "ops:Label_File_Info.ops:creation_date_time";
@@ -591,7 +591,9 @@ public class ValidateLauncher {
     } catch (Exception e) {
       e.printStackTrace();
     } finally {
-      reader.close();
+      if (reader != null) {
+        reader.close();
+      }
     }
 
     try {

--- a/src/main/java/gov/nasa/pds/validate/util/ToolInfo.java
+++ b/src/main/java/gov/nasa/pds/validate/util/ToolInfo.java
@@ -100,4 +100,8 @@ public class ToolInfo {
   public static String getQuery() {
     return props.getProperty(QUERY);
   }
+
+  public static void setProperty(String key, String value) {
+    props.setProperty(key, value);
+  }
 }

--- a/src/main/resources/validate.properties
+++ b/src/main/resources/validate.properties
@@ -37,4 +37,4 @@ validate.copyright=\nCopyright 2019, by the California Institute of Technology (
 validate.search_url=https://pds.nasa.gov/api
 validate.output_file_name=registered_context_products.json
 validate.endpoint=search/1/products
-validate.query=(product_class eq "Product_Context" and pds:Resource.pds:type ne "Information.Science_Portal" and pds:Resource.pds:type ne "Information.Investigation")
+validate.query=(product_class eq "Product_Context")

--- a/src/test/java/gov/nasa/pds/validate/GetLatestJsonContextTest.java
+++ b/src/test/java/gov/nasa/pds/validate/GetLatestJsonContextTest.java
@@ -1,0 +1,218 @@
+// Copyright © 2026, California Institute of Technology ("Caltech").
+// U.S. Government sponsorship acknowledged.
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// • Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+// • Redistributions must reproduce the above copyright notice, this list of
+//   conditions and the following disclaimer in the documentation and/or other
+//   materials provided with the distribution.
+// • Neither the name of Caltech nor its operating division, the Jet Propulsion
+//   Laboratory, nor the names of its contributors may be used to endorse or
+//   promote products derived from this software without specific prior written
+//   permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+package gov.nasa.pds.validate;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.sun.net.httpserver.HttpServer;
+import gov.nasa.pds.validate.constants.TestConstants;
+import gov.nasa.pds.validate.util.ToolInfo;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for ValidateLauncher.getLatestJsonContext() — the registry update path
+ * triggered by the {@code -u} / {@code --latest-json-file} flag.
+ *
+ * <p>Uses a JDK built-in {@link HttpServer} as a local mock so no network
+ * access is required.
+ *
+ * <p>Relates to GitHub issue NASA-PDS/validate#1659.
+ */
+class GetLatestJsonContextTest {
+
+  private static final String SORT_KEY = "ops:Label_File_Info.ops:creation_date_time";
+
+  private HttpServer server;
+  private File outputFile;
+  private String originalSearchUrl;
+
+  @BeforeEach
+  void setUp() throws Exception {
+    originalSearchUrl = ToolInfo.getSearchURL();
+    System.setProperty("resources.home", TestConstants.RESOURCES_DIR);
+    outputFile = File.createTempFile("registered_context_products_test_", ".json");
+    outputFile.deleteOnExit();
+  }
+
+  @AfterEach
+  void tearDown() throws Exception {
+    if (server != null) {
+      server.stop(0);
+    }
+    ToolInfo.setProperty(ToolInfo.SEARCH_URL, originalSearchUrl);
+  }
+
+  /**
+   * Happy path: mock server returns 2 context products in a single page.
+   * Verifies the output file is written and contains the expected number of entries.
+   */
+  @Test
+  void testSuccessfulSinglePageDownload() throws Exception {
+    String responseBody = buildPageResponse(2, 2, "2024-01-01T00:00:00Z");
+    server = startMockServer(200, responseBody);
+    int port = server.getAddress().getPort();
+    ToolInfo.setProperty(ToolInfo.SEARCH_URL, "http://localhost:" + port);
+
+    ValidateLauncher launcher = new ValidateLauncher();
+    launcher.setReportStyle("full");
+    launcher.setupReport(new String[0]);
+    launcher.setRegisteredProductsFile(outputFile);
+
+    launcher.getLatestJsonContext();
+
+    assertTrue(outputFile.exists(), "Output file should be written");
+    assertTrue(outputFile.length() > 0, "Output file should not be empty");
+
+    String content = new String(Files.readAllBytes(outputFile.toPath()), StandardCharsets.UTF_8);
+    JsonObject json = new Gson().fromJson(content, JsonObject.class);
+    assertTrue(json.has("Product_Context"), "Output JSON should contain Product_Context array");
+    JsonArray products = json.getAsJsonArray("Product_Context");
+    assertEquals(2, products.size(), "Should have written 2 context products");
+  }
+
+  /**
+   * Pagination: mock server reports total=3 but serves 2 per page; the second page
+   * supplies the remaining 1. Verifies the do/while loop pages correctly.
+   */
+  @Test
+  void testPaginatedDownload() throws Exception {
+    // Page 1: hits=3, returns 2 items
+    String page1 = buildPageResponse(3, 2, "2024-01-02T00:00:00Z");
+    // Page 2: hits=3, returns 1 item (loop terminates when contexts.size() >= total)
+    String page2 = buildPageResponse(3, 1, "2024-01-03T00:00:00Z");
+
+    int[] requestCount = {0};
+    server = HttpServer.create(new InetSocketAddress(0), 0);
+    server.createContext("/" + ToolInfo.getEndpoint(), exchange -> {
+      String body = requestCount[0]++ == 0 ? page1 : page2;
+      byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
+      exchange.sendResponseHeaders(200, bytes.length);
+      try (OutputStream os = exchange.getResponseBody()) {
+        os.write(bytes);
+      }
+    });
+    server.start();
+
+    int port = server.getAddress().getPort();
+    ToolInfo.setProperty(ToolInfo.SEARCH_URL, "http://localhost:" + port);
+
+    ValidateLauncher launcher = new ValidateLauncher();
+    launcher.setReportStyle("full");
+    launcher.setupReport(new String[0]);
+    launcher.setRegisteredProductsFile(outputFile);
+
+    launcher.getLatestJsonContext();
+
+    assertEquals(2, requestCount[0], "Should have made exactly 2 HTTP requests for pagination");
+    assertTrue(outputFile.exists(), "Output file should be written after paginated download");
+  }
+
+  /**
+   * Error path: mock server returns HTTP 400. Verifies the method catches the
+   * IOException and does not throw, leaving the output file unwritten or unchanged.
+   */
+  @Test
+  void testHttp400DoesNotThrow() throws Exception {
+    server = startMockServer(400, "Bad Request");
+    int port = server.getAddress().getPort();
+    ToolInfo.setProperty(ToolInfo.SEARCH_URL, "http://localhost:" + port);
+
+    File originalFile = outputFile;
+    // Remove the file so we can assert it was NOT written on error
+    Files.deleteIfExists(originalFile.toPath());
+
+    ValidateLauncher launcher = new ValidateLauncher();
+    launcher.setReportStyle("full");
+    launcher.setupReport(new String[0]);
+    launcher.setRegisteredProductsFile(originalFile);
+
+    // Should not throw
+    launcher.getLatestJsonContext();
+
+    assertFalse(originalFile.exists(), "Output file should not be written when registry returns 400");
+  }
+
+  // --- helpers ---
+
+  private HttpServer startMockServer(int statusCode, String body) throws IOException {
+    HttpServer s = HttpServer.create(new InetSocketAddress(0), 0);
+    s.createContext("/" + ToolInfo.getEndpoint(), exchange -> {
+      byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
+      exchange.sendResponseHeaders(statusCode, bytes.length);
+      try (OutputStream os = exchange.getResponseBody()) {
+        os.write(bytes);
+      }
+    });
+    s.start();
+    return s;
+  }
+
+  /**
+   * Builds a minimal PDS Search API JSON response with {@code count} synthetic
+   * Product_Context entries, reporting {@code total} hits.
+   *
+   * <p>The {@code properties} object mirrors the actual API shape that
+   * {@code parseJsonObjectWriteTofile} reads: {@code lidvid}, {@code title},
+   * the sort key, and at least one typed field.
+   */
+  private String buildPageResponse(int total, int count, String creationDate) {
+    StringBuilder data = new StringBuilder("[");
+    for (int i = 0; i < count; i++) {
+      if (i > 0) data.append(",");
+      String lidvid = "urn:nasa:pds:context:target:test.target_" + i + "::1.0";
+      data.append("{")
+          .append("\"id\":\"").append(lidvid).append("\",")
+          .append("\"type\":\"Product_Context\",")
+          .append("\"properties\":{")
+          .append("\"lidvid\":[\"").append(lidvid).append("\"],")
+          .append("\"title\":[\"Test Target ").append(i).append("\"],")
+          .append("\"").append(SORT_KEY).append("\":[\"").append(creationDate).append("\"],")
+          .append("\"pds:Target.pds:type\":[\"Asteroid\"]")
+          .append("}}");
+    }
+    data.append("]");
+
+    return "{\"summary\":{\"hits\":" + total + "},\"data\":" + data + "}";
+  }
+}

--- a/src/test/java/gov/nasa/pds/validate/GetLatestJsonContextTest.java
+++ b/src/test/java/gov/nasa/pds/validate/GetLatestJsonContextTest.java
@@ -1,33 +1,3 @@
-// Copyright © 2026, California Institute of Technology ("Caltech").
-// U.S. Government sponsorship acknowledged.
-//
-// All rights reserved.
-//
-// Redistribution and use in source and binary forms, with or without
-// modification, are permitted provided that the following conditions are met:
-//
-// • Redistributions of source code must retain the above copyright notice,
-//   this list of conditions and the following disclaimer.
-// • Redistributions must reproduce the above copyright notice, this list of
-//   conditions and the following disclaimer in the documentation and/or other
-//   materials provided with the distribution.
-// • Neither the name of Caltech nor its operating division, the Jet Propulsion
-//   Laboratory, nor the names of its contributors may be used to endorse or
-//   promote products derived from this software without specific prior written
-//   permission.
-//
-// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
-// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
-// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
-// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
-// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
-// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
-// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
-// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-// POSSIBILITY OF SUCH DAMAGE.
-
 package gov.nasa.pds.validate;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;

